src/extensions/scratch3_microbit/index.js

```diff
@@ -0,0 +1,676 @@
+                    TiltDirection.RIGHT, TiltDirection.ANY
+                ],
+                touchPins: ['0', '1', '2']
+            }
+        };
+    }
+
+    /**
+     * Use the Device Manager client to attempt to connect to a MicroBit device.
+     */
+    connect () {
+        this._device = new MicroBit(null, this.runtime);
+        window.addEventListener('message', event => {
+            if (event.data.type === 'data') {
+                this._device._processData(new Uint8Array(event.data.buffer));
+            }
+        }, false);
+        /*
+         * if (this._device || this._finder) {
+         *     return;
+         * }
+         * const deviceManager = this.runtime.ioDevices.deviceManager;
+         * const finder = this._finder =
+         *     deviceManager.searchAndConnect(Scratch3MicroBitBlocks.EXTENSION_NAME, MicroBit.DEVICE_TYPE, DEV_SPEC);
+         *
+         * this._finder.promise.then(
+         *     socket => {
+         *         if (this._finder === finder) {
+         *             this._finder = null;
+         *             this._device = new MicroBit(socket, this.runtime);
+         *         } else {
+         *             log.warn('Ignoring success from stale MicroBit connection attempt');
+         *         }
+         *     },
+         *     reason => {
+         *         if (this._finder === finder) {
+         *             this._finder = null;
+         *             log.warn(`MicroBit connection failed: ${reason}`);
+         *         } else {
+         *             log.warn('Ignoring failure from stale MicroBit connection attempt');
+         *         }
+         *     });
+         */
+    }
+
+    /**
+     * Test whether the A or B button is pressed
+     * @param {object} args - the block's arguments.
+     * @return {boolean} - true if the button is pressed.
+     */
+    whenButtonPressed (args) {
+        if (args.BTN === 'any') {
+            return this._device.buttonA | this._device.buttonB;
+        } else if (args.BTN === 'A') {
+            return this._device.buttonA;
+        } else if (args.BTN === 'B') {
+            return this._device.buttonB;
+        }
+        return false;
+    }
+
+    /**
+     * Test whether the micro:bit is moving
+     * @return {boolean} - true if the micro:bit is moving.
+     */
+    whenMoved () {
+        return (this._device.gestureState >> 2) & 1;
+    }
+
+    /**
+     * Test whether the micro:bit is shaken
+     * @return {boolean} - true if the micro:bit is shaken.
+     */
+    whenShaken () {
+        return this._device.gestureState & 1;
+    }
+
+    /**
+     * Test whether the micro:bit is free falling
+     * @return {boolean} - true if the micro:bit is free falling.
+     */
+    whenJumped () {
+        return (this._device.gestureState >> 1) & 1;
+    }
+
+    /**
+     * Display text on the 5x5 LED matrix.
+     * @param {object} args - the block's arguments.
+     * Note the limit is 20 characters
```

Comment:
Should we trim the string down to 20 characters here, to avoid sending e.g. the entire lyrics of "All Star" by Smash Mouth over the socket?

Reply:
**SOMEBODY ONCE TOLD ME** ok I'll stop

Reply:
This is done automatically because of the maximum bluetooth characteristic length, but it's good practice to catch this in the extension too.